hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieFileIndex.scala

```diff
@@ -349,10 +350,15 @@ case class HoodieFileIndex(spark: SparkSession,
       Option.empty
     } else if (recordKeys.nonEmpty) {
       Option.apply(recordLevelIndex.getCandidateFiles(getAllFiles(), recordKeys))
-    } else if (functionalIndex.isIndexAvailable && !queryFilters.isEmpty) {
+    } else if (functionalIndex.isIndexAvailable && queryFilters.nonEmpty && functionalIndex.extractSparkFunctionNames(queryFilters).nonEmpty) {
+      val functionToColumnNames = functionalIndex.extractSparkFunctionNames(queryFilters)
+      // Currently, only one functional index in the query is supported. HUDI-7620 for supporting multiple functions.
+      checkState(functionToColumnNames.size == 1, "Currently, only one function with functional index in the query is supported")
+      val (indexFunction, targetColumnName) = functionToColumnNames.head
+      val partitionOption = functionalIndex.getFunctionalIndexPartition(indexFunction, targetColumnName)
       val prunedFileNames = getPrunedFileNames(prunedPartitionsAndFileSlices)
       val shouldReadInMemory = functionalIndex.shouldReadInMemory(this, queryReferencedColumns)
-      val indexDf = functionalIndex.loadFunctionalIndexDataFrame("", shouldReadInMemory)
+      val indexDf = functionalIndex.loadFunctionalIndexDataFrame(partitionOption.get, shouldReadInMemory)
```

Comment:
`partitionOption` could be empty?

Reply:
Handled the case. Now, functional index partition option will get lazily initialized and the whole if block is called if the option is nonEmpty.